docs/VirtualMachine.md

```diff
@@ -0,0 +1,160 @@
+
+#### Stream
+
+A stream is an unbounded set of tables.
+Each table must have a different group key.
+
+#### Input
+
+An input is a source of a stream.
+
+Inputs are reprented as functions that produce a stream.
+
+#### Output
+
+An output consumes a stream and send the data to an external system.
+
+Output are reprented as functions that consume a stream and produce a stream.
+Outputs are transparent proxies, passing their input data downstream unmodified.
+As a result outputs themselves have no effect on the Flux program, in other words they only have side effects.
+
+#### Transformation
+
+A transformation is not a data type but rather an operation defined on a stream.
+A transformation consumes a stream of tables and produces a new stream of tables.
+
+Transformations may make use of a data cache to allow for multiple edits to a table before materializing the table into the new stream.
+
+Transformations are reprented as functions that consume a stream and produce a stream.
+
+#### Data cache
+
+A data cache is a set of tables that have yet to be materialized.
+Tables in a data cache may receive edits.
+The cache identifies each table by its group key.
+
+#### Triggers
+
+A trigger indicates that a table in the data cache should be materialized and delivered downstream.
+
+A trigger can fire multiple times until it is finished.
+Once a trigger is finished it will never fire again.
+
+A data cache uses triggers to know when it can evit and materialize tables.
+
+Triggers fire based on events.
+For example a trigger may fire after receiving an update watermark event.
+See the Flux spec for complete details on events available to triggers.
+
+#### Message bus
+
+The message bus is a queue where trigger events and data messages can be placed.
+Each transformation has its own message bus.
+A transformation consumes messages off its bus and processes them.
+
+### Data flow
+
+With the various data types and operations defined lets be explicit about how data flows through a Flux program.
+
+A Flux program describes a graph through which data flows.
+This graph may have complex data dependencies between transformations.
+Sources produce an inital stream of tables, each table is placed on the message bus for the downstream transformations of the source.
+Each transformation is linked to its downstream transformations via the message bus.
+Sources produces trigger events about the stream which are also placed on the message bus.
+The trigger events cause tables to be materialized from the data cache, if it exists. 
+The materialized tables are placed on the message bus of the downstream transformations.
+If a transformation does not need to make edits to a table it does not need to use the data cache but can simply place its generated tables directly on the downstream message bus.
+
+## Compiler
+
+The compiler is responsible for transforming a semantic graph into an IR.
```

Comment:
Would it make sense in the future to have the compiler written in Flux? I'm thinking of when we want to create Flux implementations in other languages. Then we'd only need to write the parser and the VM in the target language.

Reply:
The VM we are talking about building will not be capable of implementing a compiler, therefore Flux will not be a self-hostable language. That said we can revisit that design later as any direction towards a VM based implementation of Flux allows for a future self-hostable language.

To be specific as to some things that are missing from the Flux and the VM that would mean writing the compiler in Flux is difficult.  (Not an exhaustive list, just some initial thoughts)

* No memory management, we do not want to write a garbage collector or otherwise manage memory. The only memory management that is going to happen is around tables as units of memory which means that  queries can be well controlled and isolated, but arbitrary code will not. I expect the predicate and map functions will just have one memory free at the end of their execution. 
* Flux's type system is  too small to easily represent things like an AST. Flux has no named types for example.
* Similar to the type system Flux has only limited dynamic data structures, i.e a tables and streams. Objects, arrays and primitive values are all static, meaning they cannot grow. These dynamic data structures would be needed to write a compiler in Flux.


TL;DR Flux is a small language, adding the features to it to make it self-host the language would greatly increase the scope of the language.  That said moving towards a VM moves use towards being able to self-host Flux in the future.